_posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md

```diff
@@ -0,0 +1,549 @@
+The contract between services is set.
+
+> API first development allows teams to develop separately against a common interface, the API.
+
+Now that we understand the importance of and value of API first design let's see how the Swagger/OpenAPI spec can help you with that.
+
+## Top-down vs bottom-up
+> API first development implies a top-down approach to build your API.
+
+Basically there are two approaches:
+* Top-down aka Design First
+* Bottom-up aka Code First
+
+To quote `Swagger.io`:
+> Design First: The plan is converted to a human and machine readable contract, such as a Swagger document, from which the code is built.  
+
+> Code First: Based on the business plan, API is directly coded, from which a human or machine readable document, such as a Swagger document can be generated.
+
+In this blog post I am using the Top-down Design First approach to facilitate API first development.
+In the last paragraph of this blog I'll briefly show an example of a Code First approach with `Springfox`.
+
+# API design: an example
+Suppose that we, ***Ordina*** are hosting a conference where multiple technical and agile sessions will be given.
+Users can check session information and register for sessions.
+The backend service is accessed by a web application and two mobile apps.
+
+<div style="text-align: center;">
+  <img src="/img/2019-09-12-API-first-development-with-OpenAPI-or-Swagger/design.png" width="40%" height="40%" target="_blank">
+</div>
+
+The applications are created by different teams and they all embrace the **API first** approach.
+
+They read this blog and realised that by agreeing on the common interface first they could develop separately without impacting each other.
+So hooray for API first development!
+
+# API first development with OpenAPI/Swagger
+Let's continue with creating the backend application.
+
+The functional analysts and a couple of developers of the team are sitting together to agree on how the API should be defined.
+* A client should be able to fetch all sessions via the API
+* A client should be able to create a new session via the API
+
+This is a crucial part of the API first mindset.
+We need to clearly define and communicate the API before starting to implement.
+Designing an API is easier when you can visualise the API.
+Let's bring in the OpenAPI spec.
+
+> The OpenAPI specification allows you to define your API in a descriptive language (JSON or Yaml) and nicely visualise it
+
+Note: If I speak about OAS (OpenAPI Specification) instead of OpenAPI/Swagger specification.  
+Let's now use OAS to help us with our API First approach and design our API.
+
+Let me introduce you to [https://editor.swagger.io](https://editor.swagger.io){:target="_blank"}.
+A portal to visualise your OAS.
+Easy to use and offering all the functionality we need for this example.
+
+I'll keep it simple and create the OAS for the API exposing the endpoint where consumers of the API can fetch all sessions of the conference.
```

Comment:
I'll keep it simple, we will create the OAS for exposing the endpoint to let consumers fetch the sessions of the conference